### PR TITLE
Changing way to apply auto-style values

### PR DESCRIPTION
**Pull request by xavijam**
Related ticket: https://github.com/CartoDB/deep-insights.js/issues/557.

Instead of keep going with the previous way of setting auto-style values, we have changed several things in order to improve what we had.

- Now we use PostCSS parser to edit/remove/create properties from the current CartoCSS (thanks to @rochoa for the idea).
- We just replace or remove CartoCSS declarations, we don't add any other new symbolizers (like `mapnik::geometry_type`).
- Removed unnecessary and fake code related to auto-style.

## Acceptance
Using a polygon dataset as example, add a histogram widget.

### 1
- Set a normal CartoCSS
```css
#layer {
  polygon-fill: #343943;
  polygon-opacity: 0.9;
}
```
- Apply auto-style and check everything is ok, it should be:

```css
#layer {
  polygon-fill: ramp([column_name], cartocolor(ramp, x), whatever);
  polygon-opacity: 0.9;
}
```

### 2
- Set a CartoCSS with `::outline`
```css
#layer {
  polygon-fill: #343943;
  polygon-opacity: 0.9;
  ::outline {
  	line-width: 1;
  	line-color: #FFF;
  	line-opacity: 0.5;
  }
}
```
- Apply auto-style and check everything is ok, it should be:

```css
#layer {
  polygon-fill: ramp([a03_bloc_1], cartocolor(one, X), whatever);
  polygon-opacity: 0.9;
  ::outline {
  	line-width: 1;
  	line-color: #FFF;
  	line-opacity: 0.5;
  }
}
```

### 3
- Set a CartoCSS with conditional styles
```css
#layer {
  polygon-fill: #343943;
  polygon-opacity: 0.9;
  [column_name>10] {
  	polygon-fill: red;
        polygon-opacity: 1;
  }
  ::outline {
  	line-width: 1;
  	line-color: #FFF;
  	line-opacity: 0.5;
  }
}
```
- Apply auto-style and check everything is ok, it should be:

```css
#layer {
  polygon-fill: ramp([column_name], cartocolor(one, X), jenks);
  polygon-opacity: 0.9;
  [column_name>10] {
    polygon-opacity: 1;
  }
  ::outline {
  	line-width: 1;
  	line-color: #FFF;
  	line-opacity: 0.5;
  }
}
```

### 4
- Set a CartoCSS with conditional styles and new symbolizer.
```css
#layer {
  polygon-fill: #343943;
  polygon-opacity: 0.9;
  [column_name>10] {
  	polygon-fill: red;
  }
  ::outline {
  	line-width: 1;
  	line-color: #FFF;
  	line-opacity: 0.5;
  }
}
#layer['mapnik::geometry_type'=3] {
  polygon-fill: blue;
  polygon-opacity: 0.3;
}
```
- Apply auto-style and check everything is ok, it should be:

```css
#layer {
  polygon-fill: ramp([column_name], cartocolor(one, x), whatever);
  polygon-opacity: 0.9;
  [column_name>10] {
  }
  ::outline {
  	line-width: 1;
  	line-color: #FFF;
  	line-opacity: 0.5;
  }
}
#layer['mapnik::geometry_type'=3] {
  polygon-opacity: 0.3;
}
```

cc @rochoa @nobuti

### CODE REVIEW

File: src/widgets/auto-style/style-utils.js [hunk 1]
```diff
@@ -1,67 +1,76 @@
...
+  });
 
-  return exist ? cartocss.replace('{', '{ ' + flag) : cartocss;
+  return propertyIncluded;
 }
 
-function removeAttr (cartocss, attr) {
-  return cartocss.replace(getAttrRegex(attr, true), '');
+function isMapnikGeometrySelectorRule (node) {
+  return node.type === 'rule' && node.selector.search('mapnik::geometry_type') !== -1;
```
Comment (alonsogarciapablo) [inline src/widgets/auto-style/style-utils.js]: I would use `isSelectorRule` here too:

```
return isSelectorRule(node) && node.selector.search('mapnik::geometry_type') !== -1;
```

File: src/widgets/auto-style/style-utils.js [hunk 2]
```diff
@@ -1,67 +1,76 @@
...
-function removeEmptyLayer (cartocss) {
-  return cartocss.replace(/[^;}{]*{((\s|\n)*?)}/g, '');
-}
-
-function setFlagInCartocss (cartocss, attr, flag) {
-  var exist = cartocss.search(getAttrRegex(attr, false)) >= 0;
+  root.walk(function (node) {
+    var parentNode = node.parent;
+    if (node.type === 'decl' && node.prop === attr) {
+      if (!isSelectorRule(parentNode) || isMapnikGeometrySelectorRule(parentNode)) {
```
Comment (alonsogarciapablo) [inline src/widgets/auto-style/style-utils.js]: Is not very clear by looking into this conditional when a property is included...

File: spec/widgets/auto-style/style-utils.spec.js [hunk 3]
```diff
@@ -5,39 +5,49 @@ describe('src/widgets/auto-style/style-utils', function () {
...
     });
 
     it('should change style, remove duplicated and remove empty braces', function () {
       var cartocss = '#layer { marker-line-color: white; marker-fill: #adadad; [me>gasol] { marker-fill: #fafafa; }}';
       var attr = 'marker-fill';
-      var newStyle = 'marker-fill: blue;';
+      var newStyle = 'blue';
 
-      var expected = "#layer ['mapnik::geometry_type'=1] { marker-fill: blue;  } #layer { marker-line-color: white;}";
+      var expected = '#layer { marker-line-color: white; marker-fill: blue; [me>gasol] { }}';
```
Comment (xavijam) [inline spec/widgets/auto-style/style-utils.spec.js]: Fixing it!

File: src/widgets/auto-style/style-utils.js (hunk 2 above)
Comment (xavijam) [inline src/widgets/auto-style/style-utils.js]: Ahh, now I remember why I included this. We should consider that the property is defined if it is inside a mapnik geometry selector. If it is not inside a selector-rule (considitional) or if it is inside a selector rule but it is defined by the mapnik-geometry thing.

Comment (xavijam) [inline src/widgets/auto-style/style-utils.js]: Adding tests and applying your feedback about isSelectorRule inside method.

**Comment by xavijam** [comment]
Reviewed @alonsogarciapablo! 💋

File: src/widgets/auto-style/style-utils.js (hunk 2 above)
Comment (alonsogarciapablo) [inline src/widgets/auto-style/style-utils.js]: :+1:

**Comment by xavijam** [comment]
Planning to merge it tomorrow morning @alonsogarciapablo 💋

**Comment by alonsogarciapablo** [comment]
:+1:

**Pull request merged**